### PR TITLE
Handle entity delete events, i.e. when a repo is deleted

### PR DESCRIPTION
**Pull request by rdimitrov**
# Summary

The following PR:
* Attaches the webhook action event to the watermill message we publish
* Adds a new reconciler for processing entity `delete` events. Currently it has meaning only for artifact entities, but I've left it as a switch so if we decide to handle other events which mean a similar action for other entities we can easily extend it., i.e. closed for pull requests.
* Renamed and put all watermill event constants to be part of `./internal/events/constants.go`
* Added constants for webhook action events, i.e. "deleted", "opened", etc.
* Added handling for the use case when a webhook is deleted for a registered repo. 

Fixes https://github.com/stacklok/minder/issues/2927

## Change Type

***Mark the type of change your PR introduces:***

- [x] Bug fix (resolves an issue without affecting existing features)
- [ ] Feature (adds new functionality without breaking changes)
- [ ] Breaking change (may impact existing functionalities or require documentation updates)
- [ ] Documentation (updates or additions to documentation)
- [ ] Refactoring or test improvements (no bug fixes or new functionality)

# Testing

- [x] I'd like to add a smoke test for this so we merge them in parallel - https://github.com/stacklok/minder-smoke-tests/pull/93

# Review Checklist:

- [x] Reviewed my own code for quality and clarity.
- [ ] Added comments to complex or tricky code sections.
- [ ] Updated any affected documentation.
- [ ] Included tests that validate the fix or feature.
- [ ] Checked that related changes are merged.

### CODE REVIEW

File: internal/engine/executor.go [hunk 1]
```diff
@@ -209,14 +208,55 @@ func (e *Executor) prepAndEvalEntityEvent(ctx context.Context, inf *entities.Ent
...
 		Provider: Provider{
 			Name: inf.Provider,
 		},
 	}
 
-	return e.evalEntityEvent(ctx, inf, ectx, cli)
+	// Channel the event based on the webhook action - create, update, delete, published, etc.
+	switch inf.ActionEvent {
+	case "deleted":
+		return e.deleteEntityEvent(ctx, inf, ectx)
```
Comment (JAORMX) [inline internal/engine/executor.go:221]: Uhm... I was thinking we'd have a reconciler dedicated towards delete events, and keep the executor just executing policy.

Comment (rdimitrov) [reply internal/engine/executor.go:221]: That's a fair point. I started writing it, but then I felt it's not worth to add so much code just for this use case as it turned out it is not really useful for anything else yet other than repositories, i.e. artifacts and pull requests don't have events similar to `delete`. If you say it'd be better for future where we have different providers and such I can update it, no problem 👍

Comment (JAORMX) [reply internal/engine/executor.go:221]: Let's keep the executor just executing. Having a general deletion reconciler will pay off on the future.

Comment (rdimitrov) [reply internal/engine/executor.go:221]: Alright, I'll rework the PR then 👍

File: internal/engine/entities/entity_event.go [hunk 2]
```diff
@@ -287,6 +298,13 @@ func (eiw *EntityInfoWrapper) withProviderFromMessage(msg *message.Message) erro
 	return nil
 }
 
+//nolint:unparam
```
Comment (dmjb) [inline internal/engine/entities/entity_event.go:301]: Why is this needed? It looks like the argument and the method receiver are both used.

Comment (rdimitrov) [reply internal/engine/entities/entity_event.go]: Yeah, it was because we always returned nil as an error. I had as I thought to enforce having it but forgot to remove it. Thanks for noticing this!

File: internal/engine/executor.go (hunk 1 above)
Comment (rdimitrov) [reply internal/engine/executor.go]: Updated the PR by adding a new reconciler and also moved queue topics to live under `events/constants.go` so it's more consistent.

**Comment by rdimitrov** [comment]
Opened a PR with the smoke test for this:
* https://github.com/stacklok/minder-smoke-tests/pull/93

**Review by evankanderson — COMMENTED** [review]
This generally looks good -- once concern about the return value of `handleEntityDeleteEvent`, and some comments about combining refactoring with new code.  😁

File: internal/reconcilers/repository.go [hunk 3]
```diff
@@ -37,11 +36,6 @@ import (
 	pb "github.com/stacklok/minder/pkg/api/protobuf/go/minder/v1"
 )
 
-var (
-	// ArtifactTypeContainerRetentionPeriod represents the retention period for container artifacts
-	ArtifactTypeContainerRetentionPeriod = time.Now().AddDate(0, -6, 0)
-)
-
```
[the remark this answers is not in the document]
Comment (rdimitrov) [reply internal/reconcilers/repository.go:44]: yes, it was a leftover for when we stopped storing the artifact vesions 👍

File: internal/reconcilers/entity_delete.go [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+		// will automatically remove them when the repository is deleted.
+		if err := r.store.DeleteRepository(ctx, repoID); err != nil {
+			return fmt.Errorf("error deleting repository from DB: %w", err)
+		}
+		minderlogger.BusinessRecord(ctx).Repository = repoID
+		return nil
+	default:
+		err := fmt.Errorf("unsupported entity delete event for: %s", inf.Type)
+		l.Err(err).Msg("error handling entity delete event")
+		return err
```
[the remark this answers is not in the document]
Comment (rdimitrov) [reply internal/reconcilers/entity_delete.go:62]: good point 👍 I'd say it still make sense to log as error though right? Just so we know something was not as expected

File: internal/controlplane/handlers_githubwebhooks_test.go
[the remark this answers is not in the document]
Comment (rdimitrov) [reply internal/controlplane/handlers_githubwebhooks_test.go:1]: yes, there's an opened PR with a smoke test for this 👍

[the remark this answers is not in the document]
Comment (evankanderson) [reply internal/controlplane/handlers_githubwebhooks_test.go:1]: I'd meant a unit test, but that works.

File: internal/reconcilers/entity_delete.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (evankanderson) [reply internal/reconcilers/entity_delete.go:62]: Yes on logging the error, particularly if we don't return it.  My feeling would be that we should _either_ handle the error _XOR_ return an error.  Logging the error is one way of handling it.

[the remark this answers is not in the document]
Comment (rdimitrov) [reply internal/reconcilers/entity_delete.go]: Fixed it 👍

File: internal/reconcilers/entity_delete.go [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+	if err != nil {
+		return fmt.Errorf("error unmarshalling payload: %w", err)
+	}
+
+	l := zerolog.Ctx(ctx).With().
+		Str("provider", inf.Provider).
+		Str("project_id", inf.ProjectID.String()).
+		Str("entity_type", inf.Type.ToString()).
+		Str("action", inf.ActionEvent).
+		Logger()
```
[the remark this answers is not in the document]
Comment (rdimitrov) [reply internal/reconcilers/entity_delete.go:42]: Sounds good, let's add this in a separate PR 👍

**Review by JAORMX — APPROVED** [review]
